Functional Stepper Motor Driver Hardware Interface Implementation

corbenkolaski opened the pull request:
The stepper motor driver can successfully launch, as well as retrieve and process commands from the action server. However, the implementation may still require alterations or additional functionality to be added in the future. This branch contains the changes introduced in PR #6. Based on the comments in PR #6 and internal discussions amongst the team, the following changes were made:
- fixed issues in the ros2_control.xacro and verified on_init parsing with various configurations.
- Added devices and GPIO passthrough to the prod docker file and also fixed linking issues
- set `async`=true in the interface ros2_control tags
- minor refactoring to the CL42T interface implementation.
- Over-rode on_activate and on_deactive methods in the interface
- Improved the xacro definitions to more accurately reflect the physical setup of the robot

corbenkolaski committed:
corrected name space to match package name

corbenkolaski committed:
Corrected the ros2_control tag names

I assigned unique tags to each interface instance, and I also checked the on_init validations worked in various configurations.

corbenkolaski committed:
Added on_activate and on_deactivate implementations

Currently this just sets the direction pin and waits the minimum setup time on activate. I also changed the name space  for better encapsulation within the package

corbenkolaski committed:
set the stepper interfaces to run asynchronously.

The interfaces should now run in run separate threads with the default thread priority.

corbenkolaski committed:
Refactored pulse_generation, and timing.

I changed the sleep function calls made for single use versus perioidic timing. For periodic timing such as the pulse generation, I use a single rate object each call rather then instantiating a new object each pulse  iteration. I also refactored the generate_pulses method to simply operate with the recommended duty cycle since the additional generality wasn't used and made the logic more complex. FInally, I corrected the logger access in generate_pulses which required storing the name of the hardware component as a variable.

corbenkolaski committed:
simplified logger access

Sedkian committed:
Added devices and GPIOs passthrough to docker run

Sedkian committed:
Fixed GPIOD library linking issues

Sedkian committed:
Fixed GPIO permission access error: Added user to GPIO group

Sedkian committed:
Added instructions for GPIO setup on Pi

Sedkian committed:
Updated the readme to include details for sending commands

Sedkian committed:
Added command/state Logs to the stepper HWI

corbenkolaski committed:
Improved Carriage position ranges.

Created a parameter specifying the piano length which improves readibility, and modified the carriage ranges so that the hands can no longer overlap at each end.

DobrowneyT commented on src/hardware/src/cl42t.cpp:
```diff
@@ -0,0 +1,282 @@
+        }
+
+        // Reset the state interface
+        set_state(position_state_interface_name_, 0.0);
```
This is setting the starting position to 0, even though it doesn't truly know where the carriage is. Once we have the homing routine it will just re-update the zero position?

corbenkolaski replied:
Yes, that's correct. The code right now assumes the carriage is starting at 0 so any difference in the physical location of the carriage isn't accounted for yet, but as you said, once the homing routine is implemented the state should match the physical location.

DobrowneyT reviewed (approved):
Looks good, and we know it works :)

Sedkian reviewed (approved):
Looks good!

corbenkolaski committed:
Modified the carriage range to account for hand lengths.

I updated the calculations for the carriage min and max positions and also changed the hand length to match the carriage for right now so I could validate the calculations with the physical setup. Once the actual hand is designed we can update this value.
